posts/java-enums-with-extra-visitor-please.md

```diff
@@ -0,0 +1,203 @@
+        }
+    },
+    AGRICULTURAL {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitAgricultural();
+        }
+    };
+
+    public abstract <E> E accept(AssetClassVisitor<E> visitor);
+}
+```
+
+The code is now ready to forget the switch-case and use an implementation of the `AssetClassVisitor`:
+
+```java
+public AutomatedTradingStrategy getAutomatedTradingStrategy(AssetClass assetClass) {
+    return assetClass.accept(new AssetClassVisitor<AutomatedTradingStrategy>() {
+        @Override
+        public AutomatedTradingStrategy visitMetal() {
+            return new HedgingStrategy();
+        }
+
+        @Override
+        public AutomatedTradingStrategy visitEnergy() {
+            return new SwingTradingStrategy();
+        }
+
+        @Override
+        public AutomatedTradingStrategy visitAgricultural() {
+            return new DayTradingStrategy();
+        }
+    });
+}
+```
+
+It can be observed that each value of `AssetClass` is responsible for calling the appropriate method of `AssetClassVisitor`. Values of `AssetClass` can now be ignored as the semantic is brought by the visitor. `AssetClass.AGRICULTURAL` could be renamed `AssetClass.AGRI` without changes where the business logic occurs.
+Furthermore, handling default behavior is not required anymore. Possibilities are now scoped to the one provided by the interface.
+
+# Add a new asset class
+
+The business is growing and activities are extending to livestock and meat. 
+This is as simple as adding `AssetClass.LIVESTOCK_AND_MEAT` value and updating the visitor interface.
+
+```java
+public enum AssetClass {
+    METAL {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitMetal();
+        }
+    },
+    ENERGY {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitEnergy();
+        }
+    },
+    AGRICULTURAL {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitAgricultural();
+        }
+    },
+    // The new value
+    LIVESTOCK_AND_MEAT {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitLiveStockAndMeat();
+        }
+    };
+
+    public abstract <E> E accept(AssetClassVisitor<E> visitor);
+}
+```
+
+```java
+public interface AssetClassVisitor<T> {
+    T visitMetal();
+    T visitEnergy();
+    T visitAgricultural();
+    // The new method
+    T visitLiveStockAndMeat();
+}
+```
+
+After this, the code will light up like a Christmas Tree: it does not compile anymore. The compiler should be thanked for doing such a great job! All those highlighted errors show that some part of the code is not designed to handle the new value yet. Let's correct this by using an exception: the existing features are not available yet for livestock and meat.
```

Comment:
All of those highlighted errors show that some part of the code is not designed to handle the new value yet. Let's correct this by using an exception: the existing features are not yet available for livestock and meat.